Framework/Algorithms/src/PDCalibration.cpp

```diff
@@ -0,0 +1,808 @@
+                                           const double tzero) {
+  vector<double> tofminmax(2);
+  if (difa == 0) {
+    tofminmax[0] = 0.;
+    tofminmax[1] = DBL_MAX;
+  } else {
+    double tofExtrema = tzero - difc * difc / (4 * difa);
+    if (difa < 0) {
+      tofminmax[0] = 0.;
+      tofminmax[1] = tofExtrema;
+    } else {
+      tofminmax[0] = std::max(0., tofExtrema);
+      tofminmax[1] = DBL_MAX;
+    }
+  }
+  return tofminmax;
+}
+MatrixWorkspace_sptr PDCalibration::load(const std::string filename) {
+  // TODO this assumes that all files are event-based
+  const double maxChunkSize = getProperty("MaxChunkSize");
+  const double filterBadPulses = getProperty("FilterBadPulses");
+
+  auto alg = createChildAlgorithm("LoadEventAndCompress");
+  alg->setProperty("Filename", filename);
+  alg->setProperty("MaxChunkSize", maxChunkSize);
+  alg->setProperty("FilterByTofMin", m_tofMin);
+  alg->setProperty("FilterByTofMax", m_tofMax);
+  alg->setProperty("FilterBadPulses", filterBadPulses);
+  alg->setProperty("LoadMonitors", false);
+  alg->executeAsChildAlg();
+  API::Workspace_sptr workspace = alg->getProperty("OutputWorkspace");
+
+  return boost::dynamic_pointer_cast<MatrixWorkspace>(workspace);
+}
+
+MatrixWorkspace_sptr PDCalibration::loadAndBin() {
+  m_uncalibratedWS = getProperty("UncalibratedWorkspace");
+
+  if (bool(m_uncalibratedWS)) {
+    return rebin(m_uncalibratedWS);
+  }
+
+  const std::string signalFile = getProperty("Signalfile");
+  g_log.information() << "Loading signal file \"" << signalFile << "\"\n";
+  auto signalWS = load(getProperty("SignalFile"));
+
+  const std::string backFile = getProperty("Backgroundfile");
+  if (!backFile.empty()) {
+    g_log.information() << "Loading background file \"" << backFile << "\"\n";
+    auto backWS = load(backFile);
+
+    double signalPcharge = signalWS->run().getProtonCharge();
+    double backPcharge = backWS->run().getProtonCharge();
+    backWS *= (signalPcharge / backPcharge); // scale background by charge
+
+    g_log.information("Subtracting background");
+    auto algMinus = createChildAlgorithm("Minus");
+    algMinus->setProperty("LHSWorkspace", signalWS);
+    algMinus->setProperty("RHSWorkspace", backWS);
+    algMinus->setProperty("OutputWorkspace", signalWS);
+    algMinus->setProperty("ClearRHSWorkspace", true); // only works for events
+    algMinus->executeAsChildAlg();
+    signalWS = algMinus->getProperty("OutputWorkspace");
+
+    g_log.information("Compressing data");
+    auto algCompress = createChildAlgorithm("CompressEvents");
+    algCompress->setProperty("InputWorkspace", signalWS);
+    algCompress->setProperty("OutputWorkspace", signalWS);
+    algCompress->executeAsChildAlg();
+    DataObjects::EventWorkspace_sptr compressResult =
+        algCompress->getProperty("OutputWorkspace");
+    signalWS = boost::dynamic_pointer_cast<MatrixWorkspace>(compressResult);
+  }
+
```

Comment:
Should also include `CropWorkspace` to throw out events outside of the window being used